### ISSUE TITLE
ChromeOS shutdown from powerd (or any reasons apart from shutdown command) crashes UEFI and breaks Android container for the subsequent reboot with signal 15 (AGAIN!)

### ISSUE THREAD

**Issue by khanhtranngoccva** [sebanc/brunch#164]
I have already updated UEFI from v1.5 to v1.6, but the error persists. Other Chromium distros might have tweaked powerd to fix the problem. Can you look into shutdown modules? My UEFI BIOS version is H81M-DGS R2 v1.6(latest).

**Comment by khanhtranngoccva** [comment]
I also observed that the CMOS clock was reset to UTC time.

**Comment by sebanc** [comment]
Hi,

I have added a new test release in my github "brunch-testing" repo (not this one) where I disabled CMOS update in case it could be causing an issue.

Regarding powerd, as I am not building chromeos from source it might be difficult to change things related to it but we will see that later.

Could you try the testing build, confirm that CMOS is not updated anymore and if your issue persist ?

**Comment by khanhtranngoccva** [comment]
No, it still doesn't work. ChromeOS still crashes UEFI whenever the system powers off, making the menu entry for the next reboot "??" and the second reboot will still result in a USB timeout. Rebooting with any means isn't affected.

**Comment by sebanc** [comment]
Ok, could you confirm the CMOS update part is fixed ?

Regarding powerd, I am not building chromeos from source, so I cannot modify it and it is a core part of the system so I cannot remove it. What you can do:
1) Play with powerd parameters which are located in /usr/share/power_manager/ and see if there is one which fixes your issue.
2) Try other recoveries, they may be compiled with different powerd flags.

Also, as a side note, it still looks like an efi bios issue and it is not because your efi bios is up-to-date that it means it is not bugged.

**Comment by khanhtranngoccva** [comment]
[Shutdown_failure_log.zip](https://github.com/sebanc/brunch/files/4650266/Shutdown_failure_log.zip)
Would you mind analyzing the log for errors? My ROM shuts down for the first time at 06:47PM UTC+7, so you will see the log confirming the first shutdown at 11:47AM UTC (due to CMOS reset).

**Comment by sebanc** [comment]
I had a close look at the logs, I saw nothing that would explain the menu entry issue after reboot, however there are a lot of warnings regarding the lack of battery.

**Comment by khanhtranngoccva** [comment]
What happens when there is no battery in ChromeOS? I am using a custom PC.

**Comment by sebanc** [comment]
That's a good question, i don't know actually I have only laptops. Maybe you could try a chromebox recovery image like fizz ?

Also, you could try modifying the battery related settings in /usr/share/power_manager/

**Comment by sebanc** [comment]
Also, I just had a look at the powerd source code and I have no idea of what could be causing the menu entry issue:
https://chromium.googlesource.com/chromiumos/platform2/+/master/power_manager

**Comment by khanhtranngoccva** [comment]
Well, I tried to set USB and storage settings like this and get some results:
1. Enable Intel Smart Auto: No more disk timeout error.
The Android container still crashes (even after powerwashing) and POST now hangs forever instead of the menu entry error.

**Comment by khanhtranngoccva** [comment]
2. Disable USB Legacy support: nothing changes, but the PC bricks.

**Comment by sebanc** [comment]
Is your motherboard able to boot bios/mbr devices ? Maybe if the issue is related to UEFI, bios boot would work better ?

**Comment by khanhtranngoccva** [comment]
Enabling Intel USB3.0 by default + disable boot failure guard solves the USB timeout bug, but Android container keeps failing to mount.
2020-05-20T20:27:48.602671+07:00 DEBUG arc-kmsg-logger[2157]: vdc: Waited 1030ms for vold
2020-05-20T20:27:48.602856+07:00 ERR arc-kmsg-logger[2157]: vdc: Failed: Status(-8): '0: '
2020-05-20T20:27:48.603218+07:00 INFO arc-kmsg-logger[2157]: init: Service 'exec 2 (/system/bin/vdc --wait cryptfs init_user0)' (pid 15) exited with status 25 waiting took 25.879999 seconds
2020-05-20T20:27:48.603263+07:00 ERR arc-kmsg-logger[2157]: init: Failure (reboot suppressed): init_user0_failed

**Comment by khanhtranngoccva** [comment]
Yes, but bios boot with croissant images won't fix the menu entry error, or fix the container.

**Comment by sebanc** [comment]
If you know that bios won't work, no need to try. Nice that you are making progress.

Regarding android apps, now your issue really looks like #74, have you tried waiting a bit on the login screen as suggested in this issue ?

**Comment by khanhtranngoccva** [comment]
[  107.984682] init: apex-monitor post-stop process (1392) terminated with status 1
[  111.667839] init: cros_healthd main process (1475) terminated with status 254
[  111.668744] init: cros_healthd main process ended, respawning
[  111.766844] init: cros_healthd main process (1614) terminated with status 254
[  111.767716] init: cros_healthd main process ended, respawning
[  111.848701] init: cros_healthd main process (1625) terminated with status 254
[  111.849534] init: cros_healthd main process ended, respawning
[  111.912759] init: cros_healthd main process (1632) terminated with status 254
[  111.913605] init: cros_healthd respawning too fast, stopped
[  118.933247] fscrypt: AES-256-CTS-CBC using implementation "cts(cbc(aes-generic))"
[  119.480284] fscrypt: AES-256-XTS using implementation "xts(ecb(aes-generic))"
[  154.774680] cgroup: run_oci (1361) created nested cgroup for controller "memory" which has incomplete hierarchy support. Nested cgroups may change behavior in the future.
[  154.777215] cgroup: "memory" requires setting use_hierarchy to 1 on the root
[  155.706356] EXT4-fs (sdb1): re-mounted. Opts:
[  162.020863] binder: 2313:2313 ioctl 4018620d 7ffe40b189d0 returned -22
[  183.354079] init: arc-boot-continue pre-start process (2269) terminated with status 1
[  183.683935] init: arc-kmsg-logger main process (2251) killed by TERM signal
[  183.749796] init: usb_bouncer main process (2275) terminated with status 1

**Comment by sebanc** [comment]
This one is totally normal:
[ 154.774680] cgroup: run_oci (1361) created nested cgroup for controller "memory" which has incomplete hierarchy support. Nested cgroups may change behavior in the future.

The cros_healthd issues seems more suspicious to me, maybe you can try to run:
sudo mv /etc/init/cros_healthd.conf /etc/init/cros_healthd.backup

and reboot to see if it helps.

**Comment by sebanc** [comment]
Could you try the new release I have uploaded on the "brunch-testing" branch ?

**Comment by khanhtranngoccva** [comment]
Well, I don't know what happens when the system halts and the disk has not been unmounted. Will it mean a hot plug, and what happens when a Linux Live USB gets hot-plugged?
[umount-encrypted.log](https://github.com/sebanc/brunch/files/4662256/umount-encrypted.log)

**Comment by khanhtranngoccva** [comment]
The USB timeout error has already been solved by enabling Intel USB3.0 Mode, but ChromeOS did not unmount properly and this is another issue, which may be the reason for my other issues.

**Comment by sebanc** [comment]
Indeed that's weird but I would say it is a consequence of the container not being correctly mounted in the first place.

**Comment by khanhtranngoccva** [comment]
What happens if you pull the plug of a USB when it's running Linux Live USB?

**Comment by sebanc** [comment]
You have a delay between the moment the write instruction is executed and when it is actually written on the disk. If you unplug an usb you loose the changes that have not been synced on the disk which can be quite bad or not depending on how important the data that have not been synced are.

**Comment by khanhtranngoccva** [comment]
How about pulling the plug of a PC?

**Comment by sebanc** [comment]
Exact same thing.

If you want to disable shutdown, you can try running:
`echo "exit 0" | sudo tee /usr/bin/powerd_setuid_helper`

**Comment by sebanc** [comment]
Also, I again attempted a fix in the release I have just uploaded in "brunch-testing" repository, don't hesitate to try it.

**Comment by khanhtranngoccva** [comment]
Well, I threw the ChromeOS logs into the official Google forum.

**Comment by khanhtranngoccva** [comment]
Your newer package still fails. Have you inspected arc.log carefully?

**Comment by sebanc** [comment]
Yes I did with my limited knowledge of how arc work, knowing that this part of chromeos is not open source.

**Comment by khanhtranngoccva** [comment]
NOT open source? I thought there is a page on Chromium source.

**Comment by sebanc** [comment]
The source code is not available as far as I know.

I looked at arc.log trying to find differences with other logs I have but there was nothing obvious as to why it fails (just some battery related issues which might actually be normal).

**Comment by khanhtranngoccva** [comment]
[dmesg.log](https://github.com/sebanc/brunch/files/4693974/dmesg.log)
How about this? Is there a problem with my PC or software?

**Comment by khanhtranngoccva** [comment]
Another thing is that after waiting for one minute and a half, the container boots at first login without fail, but waiting like that destroys the purpose of ChromeOS.

**Comment by khanhtranngoccva** [comment]
One more observation relating to delayed sign in: Whenever the fscrypt process runs before run_oci cgroup is created, the container shuts down, and when the fscrypt process run later than when the run_oci cgroup is created, the container starts with no fails

**Comment by khanhtranngoccva** [comment]
![image](https://user-images.githubusercontent.com/61155608/83122060-4c919200-a0fd-11ea-953e-a0dfef01c047.png)
![image](https://user-images.githubusercontent.com/61155608/83122124-616e2580-a0fd-11ea-943e-d414d311e254.png)

**Comment by sebanc** [comment]
I am not sure I understand what you are saying, correct the below if something is incorrect:
- On first boot, android apps always install without problems.
- On second boot android apps do not work if you login directly and you have to wait 1min30 for them to work. Do you have to wait 1 min 30 before login or after login ?

**Comment by khanhtranngoccva** [comment]
YES. I see that linux binder takes 3 minutes (176 seconds) from BIOS to start, while getting to log in screen takes only 2 minutes (117 seconds)

**Comment by sebanc** [comment]
Do you need to wait 1min 30 specifically at the login screen ?
Or does it work even if you login directly ?

**Comment by khanhtranngoccva** [comment]
Yes.

**Comment by sebanc** [comment]
yes to which ?

**Comment by khanhtranngoccva** [comment]
I had to wait for 1min to let the container work.

**Comment by sebanc** [comment]
I understand that you have to wait but you did not answer my questions.
Do you specifically have to wait at the login screen ?
Or does login in and waiting has the same effect ?

**Comment by khanhtranngoccva** [comment]
When I log in too soon, the device decryption starts before binder module starts which makes "binder" unusable and Android container crashes (as seen in dmesg). However, if I wait, the binder module starts before decryption and it is usable. Is it correct?

**Comment by sebanc** [comment]
The android binder is built-in the kernel, not a module so that's not it.

Moreover, since the 20200526 release, the login screen is not displayed before all modules have been loaded which has already fixed a similar issue for another user as reported in #74.

As far as I can tell, those logs seems more like the consequence of the issue than its origin.

**Comment by khanhtranngoccva** [comment]
Then why could I log in before ashmem and binder have been started?

**Comment by sebanc** [comment]
I must have missed something, what makes you think you logged in before ashmem and binder had been started ?

**Comment by khanhtranngoccva** [comment]
The system decrypts before binder sends a signal.

**Comment by sebanc** [comment]
My understanding is that the android container fails, send a bad call to the binder which in turn fails.

Here is the android binder kernel config:
CONFIG_ANDROID=y
CONFIG_ANDROID_BINDER_IPC=y
CONFIG_ANDROID_BINDER_DEVICES="binder,hwbinder,vndbinder"

**Comment by khanhtranngoccva** [comment]
Then I don't know why I had to wait in the sign in screen for a minute.

**Comment by sebanc** [comment]
I don't know also, what i can tell you for sure is that since 20200522 all modules are loaded before login screen is displayed so that's not it.

Have you tried updating chromeos to see if it helps ?

**Comment by khanhtranngoccva** [comment]
I am on latest ChromeOS.

**Comment by sebanc** [comment]
r83 ?

**Comment by khanhtranngoccva** [comment]
No. I use r80 stable.

**Comment by khanhtranngoccva** [comment]
If I log in right after this log line: [  168.224761] capability: warning: `main' uses 32-bit capabilities (legacy support in use), ARC++ works, but if I log in right before the line, it fails.

**Comment by sebanc** [comment]
Could you try editing /etc/init/arc-system-mount.conf and replacing:
"start on starting boot-services"
by
"start on started ui"

**Comment by khanhtranngoccva** [comment]
No, I still have to wait until that line in dmesg spawns to be able to log in and use ARC++.

**Comment by sebanc** [comment]
Could you revert the previous change and try editing /etc/init/ui.conf replacing:
"start on started boot-services and stopped udev-trigger"
by
"start on started boot-services and started arc-system-mount"

**Comment by khanhtranngoccva** [comment]
Still not working.
![image](https://user-images.githubusercontent.com/61155608/83247929-7e265e00-a1ce-11ea-81bd-21450eaf4e0d.png)

**Comment by sebanc** [comment]
I think this one might break boot but it is worth a try, edit

**Comment by khanhtranngoccva** [comment]
The system failed to boot completely. Can you make the process related to the 'capability: warning: `main' uses 32-bit capabilities (legacy support in use)' line start faster?

**Comment by sebanc** [comment]
That's what I am trying to do, I think it should be /etc/init/arc-system-mount.conf except that I cannot really make it start faster but I can try to delay login.

**Comment by khanhtranngoccva** [comment]
Yes, that should be acceptable.

**Comment by sebanc** [comment]
I am starting to be out of ideas, could you try editing /etc/init/arc-system-mount.conf and replace:
"start on starting boot-services"
by
"start on stopped cgroups"

**Comment by khanhtranngoccva** [comment]
Failed. Can you make the login only appear after the capability line has spawned in dmesg?

**Comment by sebanc** [comment]
Maybe it's possible but I don't know how to do that.

Could you try editing /etc/init/ui.conf replacing:
"start on started boot-services and stopped udev-trigger"
by
"start on started boot-services and started arc-boot-continue"

**Comment by khanhtranngoccva** [comment]
That also failed.

**Comment by khanhtranngoccva** [comment]
By the way, I don't know what is the 'main' process referred to in the 'capability' log line, but I'm sure it's Android related.

**Comment by sebanc** [comment]
Indeed if I am not mistaken, it is the android initramfs

**Comment by khanhtranngoccva** [comment]
What happens if it is not fully loaded before hitting the sign in button?

**Comment by khanhtranngoccva** [comment]
Let me move my Android issue into another one.

**Comment by khanhtranngoccva** [comment]
Can you try to make ui wait for exactly 1 minute before initializing?

**Comment by sebanc** [comment]
No sure if it is possible, I would not know how to do that.

Btw, why did you create another issue ?

**Comment by khanhtranngoccva** [comment]
First, it takes too long to scroll. And how can I put 'sleep 60' or 'delay 60' or 'wait 60' or whatever into ui.conf?

**Comment by khanhtranngoccva** [comment]
I've tried that method, but since android loading is after the ui service has started, delay won't work.

**Comment by sebanc** [comment]
Could you try editing /etc/init/ui.conf replacing:
"start on started boot-services and stopped udev-trigger"
by
"start on started boot-services"

and

editing /etc/init/arc-system-mount.conf and replace:
"start on starting boot-services"
by
"start on stopped udev-trigger"

**Comment by khanhtranngoccva** [comment]
The system ran normally but I'm still forced to wait.

**Comment by khanhtranngoccva** [comment]
I've updated to r83, but nothing changed yet.
![image](https://user-images.githubusercontent.com/61155608/83391733-1cf4c980-a41e-11ea-8a76-5c6efc73c523.png)

**Comment by khanhtranngoccva** [comment]
I have switched to attached HDD and it allowed faster booting and initialization. Now the delay is 30secs, but can you keep trying to make the processes run before, not after the UI?

**Comment by sebanc** [comment]
Could you try editing /etc/init/ui.conf replacing:
"start on started boot-services and stopped udev-trigger"
by
"start on started boot-services and stopped cgroups"

**Comment by khanhtranngoccva** [comment]
It still doesn't work. The arc-system-mount still starts after the login prompt gets visible.

**Comment by khanhtranngoccva** [comment]
These lines spawn repeatedly and make preloading Android significantly slower:
2020-06-13T13:05:36.883650+07:00 WARNING arc-networkd[1630]: [WARNING:crostini_service.cc(287)] Failed to get ADB sideloading status
2020-06-13T13:05:36.889098+07:00 WARNING arc-container[1697]: [stderr] ArtContainerLog:ERROR cutils-trace Error opening trace file: No such file or directory (2)
2020-06-13T13:05:42.088798+07:00 ERR arc-networkd[1630]: [ERROR:object_proxy.cc(628)] Failed to call method: org.chromium.SessionManagerInterface.QueryAdbSideload: object_path= /org/chromium/SessionManager: org.freedesktop.DBus.Error.NoReply: Did not receive a reply. Possible causes include: the remote application did not send a reply, the message bus security policy blocked the reply, the reply timeout expired, or the network connection was broken.

**Comment by sebanc** [comment]
Could you try editing /etc/init/arc-system-mount.conf replacing:
"start on starting boot-services"
by
"start on started boot-services and started arc-network-bridge"

**Comment by Skawtnyc** [comment]
ARC doesn't actually boot until after you log in, and once it does the init scripts it runs execute after ChromeOS has already run its own init scripts Unfortunately there is no "start on" descriptor to specify waiting for those ARC init scripts to finish. Your /etc/init script needs to have a sleep command in it. Make sure it looks like this:

```
start on arc-booted
stop on stopping ui or starting reboot

script

  sleep 20

  <your script commands here>

end script

```
The sleep value can be more or less, depending upon how long it usually takes for ARC to launch on your device. On my Lenovo Duet it's 20-25 seconds. On my Surface Go it's 10 seconds (SSD boots much faster).

**Comment by khanhtranngoccva** [comment]
Which is why Chrome OS is supposed to run on Chromebooks, and the devs wouldn't mind if running on a HDD causes some error. And I don't know which file your were going to edit @Skawtnyc

**Comment by Skawtnyc** [comment]
I created my own script named asc.conf and placed it in /etc/init/

**Comment by sebanc** [comment]
Have you tried the new unstable build which fixed issue #271 ?

**Comment by daemongloom** [comment]
> The sleep value can be more or less, depending upon how long it usually takes for ARC to launch on your device. On my Lenovo Duet it's 20-25 seconds. On my Surface Go it's 10 seconds (SSD boots much faster).

The general rule is to never use long sleeps. Use short one in cycle and check existance of required process.

**Comment by Skawtnyc** [comment]
> The general rule is to never use long sleeps. Use short one in cycle and check existance of required process. It increases cpu load a bit, but it reduces delay a lot.

Normally yes, but in some cases you actually need it. For instance, my script uses adb commands to change the scaling for Android apps, which the devs severely broke around version 81 when they split the browser and OS settings. Apparently there's a process within the Android container's init scripts that actually resets any saved font and display scaling for Android, which is why they don't stick past reboot. However, since Android finishes booting *after* you log in, you need a sleep command because there is no way to determine exactly when the Android boot process actually ends, and the script needs to execute after the Android init scripts are done.

In my desktop system I replaced the SATA SSD with a M.2 SSD. It's actually so fast that when I log in, my script doesn't even need a sleep delay; Android has finished booting by the time my script triggers. So, devices with EMMC and physical hard drives, as well as older SSD drives, are always going to need a custom delay to get this working.

**Comment by daemongloom** [comment]
> However, since Android finishes booting after you log in, you need a sleep command because there is no way to determine exactly when the Android boot process actually ends


But when that process starts? If it starts in the beginning and ends after logging in you should use something like that:
``  while [ `pidof PROCESS_NAME` ]
  do
    sleep 1
  done``

If not - you should also wait for "chrome" process by the same cycle. That way it your script does it work when Android has started, but you won't launch any application (with wrong DPI) before it does its work.

**Comment by Skawtnyc** [comment]
The problem is that there's no way to tell what the pid of the Android init script is because there's no way to determine the process name. The closest thing we have is the ChromeOS init script indicator "start on arc-booted" which isn't the same as "start on arc-booted-finished", which doesn't exist. I wish it did because that would make this a lot easier.

Waiting for the chrome process wouldn't work since on most systems Android is still booting while the browser is already available.

Google disabled the ability to use a modified Android image, so you can't edit the init scripts that wipe the DPI and font settings on each reboot. It's the same reason you can't root Android on ChromeOS anymore. It's too bad, because then we could just put whatever we wanted right into Android itself.

**Comment by daemongloom** [comment]
> The problem is that there's no way to tell what the pid of the Android init script is because there's no way to determine the process name. The closest thing we have is the ChromeOS init script indicator "start on arc-booted" which isn't the same as "start on arc-booted-finished", which doesn't exist. I wish it did because that would make this a lot easier.

What command exactly are you trying to do? Will it help if you use "adb wait-for-device" before your command?

**Comment by Skawtnyc** [comment]
Unfortunately no. I'm waiting for a specific Android init script to finish executing (the one that resets DPI). I don't know of any way to track that, or know when it finishes.

**Comment by khanhtranngoccva** [comment]
Appears like this: The Android container init always starts approximately 4 seconds after the login prompt aka the UI (the Upstart script for this is /etc/init/ui.conf) is launched, if you login too soon, then the container shuts down.
Therefore, can you make changes so as to:
1. Make the Android container start initializing before the UI appears.
2. Make a loop that checks $(find /run/containers/android-run_oci) or $(dmesg | grep capability) and ends when the directory is available.

**Comment by Skawtnyc** [comment]
The android container startup time is entirely dependent upon the speed of the hardware. As I mentioned, my current desktop is so fast that I don't even need a sleep delay to execute my ChromeOS init script. I'm not sure I would want to start the Android container before the UI, as there may be dependencies that need to be started first in order for Android to function (like sommelier).

I'll take a look at the android-run_oci directory, but that directory exists before the init.cheets.rc file executes. I can't rely on the existence of that directory as a trigger because it exists before the event I need to wait for. Maybe take a look at /opt/google/containers/android/rootfs/root/init.cheets.rc and see if there's anything in there that is created after the "on init" section. It has to be something that doesn't already exist and is created when the init file is executed.

**Comment by Skawtnyc** [comment]
It looks like the directory /run/containers/android-run_oci/root/dev/input is created every time the Android container boots up, and occurs after the purging of the DPI setting. I'm going to try a script that waits for it to exist and then executes. I'll let you know.

**Comment by Skawtnyc** [comment]
I have to thank you, you got me to look for something that hadn't occurred to me due to the read-only and temporary nature of the Android container. I now have a working script that instead of just sleeping for an extended period of time, the script waits for a directory to exist and then runs the commands. It works properly on every reboot now.

/etc/init/asc.conf:

start on arc-booted

script

  while ! test -f /run/containers/android-run_oci/root/dev/input

  do
    sleep 1
    sudo mkdir -p /run/arc/adb/.android
    sudo cp /home/chronos/user/Downloads/Installs/adbkey* /run/arc/adb/.android/
    adb connect 100.115.92.2
    adb wait-for-device shell wm density 450
  done

end script

**Comment by daemongloom** [comment]
> It works properly on every reboot now.

It now tries to copy file and connect in each cycle. Should be something like that:

`/etc/init/asc.conf:

start on arc-booted

script

while ! test -f /run/containers/android-run_oci/root/dev/input

do
sleep 1
done

sudo mkdir -p /run/arc/adb/.android
sudo cp /home/chronos/user/Downloads/Installs/adbkey* /run/arc/adb/.android/
adb connect 100.115.92.2
adb wait-for-device shell wm density 450

end script`

Can you check it?

**Comment by Skawtnyc** [comment]
That's how I actually had it at first, and the scaling would be proper size and then shrink back to tiny a few seconds later. I did some testing, and the file copy can be left out of it but both adb commands need to be in the while loop:

start on arc-booted

script

sudo mkdir -p /run/arc/adb/.android
sudo cp /home/chronos/user/Downloads/Installs/adbkey* /run/arc/adb/.android/

while ! test -f /run/containers/android-run_oci/root/dev/input

do
sleep 1
adb connect 100.115.92.2
adb wait-for-device shell wm density 450
done

end script

**Comment by khanhtranngoccva** [comment]
[init.zip](https://github.com/sebanc/brunch/files/4917181/init.zip)
This is a collection of init scripts from FydeOS and there's no issue with Android at all, maybe except for GApps. Can you look into it and fix my issue?

**Comment by khanhtranngoccva** [comment]
Well, I mean that you should try to make the container boot 30 seconds before the UI starts, not after. My request has nothing to do with scaling or rooting.

**Comment by Skawtnyc** [comment]
I can't "make" the container do anything. I'm not a developer, and I would have no idea where to begin with trying to launch Android before the UI is even launched. I wouldn't even want to, since the instance itself is read-only, will break if you try to extract and modify it in any way, and any customization is directly tied to the user account logging in.

Also, this isn't FydeOS.

**Comment by khanhtranngoccva** [comment]
But I would not like to wait 30 seconds to log in to Chrome OS.

**Comment by Skawtnyc** [comment]
Using the init scripts as worked out above, you can log in immediately and launch an Android app. It will take a while to load since it's still loading the Android instance, but with the correct init script syntax in place it will execute at the proper time before the Android app finally launches.

**Comment by khanhtranngoccva** [comment]
Do I really have to download adbkey?

**Comment by Skawtnyc** [comment]
> Do I really have to download adbkey?

Only if you need to send adb shell commands like I did. The Android instance deletes the adbkey files and the directory they reside in when you reboot. That's why my script recreates them.

**Comment by khanhtranngoccva** [comment]
Can I just insert the script and get the system normally, without downloading anything else from the Internet? I don't even know what adbkey is.

**Comment by Skawtnyc** [comment]
Unless you actually need the script to do something, there is no reason for putting it in. It sounds like you just want your system to boot faster, and that is a hardware issue; you would need to upgrade to make it faster. No scripts will change the speed of your system.
